core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -559,47 +571,91 @@ static ConnectionProfile resolveConnectionProfile(@Nullable ConnectionProfile co
         }
     }
 
+    protected ConnectionProfile resolveConnectionProfile(ConnectionProfile connectionProfile) {
+        return resolveConnectionProfile(connectionProfile, defaultConnectionProfile);
+    }
+
     @Override
     public final NodeChannels openConnection(DiscoveryNode node, ConnectionProfile connectionProfile) throws IOException {
         if (node == null) {
             throw new ConnectTransportException(null, "can't open connection to a null node");
         }
         boolean success = false;
         NodeChannels nodeChannels = null;
-        connectionProfile = resolveConnectionProfile(connectionProfile, defaultConnectionProfile);
+        connectionProfile = resolveConnectionProfile(connectionProfile);
         closeLock.readLock().lock(); // ensure we don't open connections while we are closing
         try {
             ensureOpen();
             try {
+                Exception connectionException = null;
+                int numConnections = connectionProfile.getNumConnections();
+                List<PlainChannelFuture<Channel>> pendingChannels = new ArrayList<>(numConnections);
+                for (int i = 0; i < numConnections; ++i) {
+                    try {
+                        PlainChannelFuture<Channel> pending = initiateChannel(node, connectionProfile.getConnectTimeout());
```

Comment:
I don't like that this returns a  future. I am almost certain it's simpler if we go with a callback here. I havent' seen a useage of future where it was actually helpful or simplifying anything.

Reply:
I guess same question as above. Are you imaging something where I pass in a CountDownLatch as a callback or a future as a callback? At some point we need to block while waiting for the connections to be completed (without changing our architecture in a pretty major way).